internal/pkg/vm/internal/vmcontext/invocation_context.go

```diff
@@ -0,0 +1,308 @@
+	targetIDAddr, ok := initView.GetIDAddressByAddress(target)
+	if ok {
+		targetActor, err := ctx.rt.state.GetActor(context.Background(), targetIDAddr)
+		if err == nil {
+			// actor found, return it and its IDAddress
+			return targetActor, targetIDAddr
+		}
+	}
+
+	// actor does not exist, create an account actor
+	// - precond: address must be a pub-key
+	// - sent init actor a msg to create the new account
+
+	if !target.IsPubKey() {
+		// Don't implicitly create an account actor for an address without an associated key.
+		panic(exitcode.ActorNotFound)
+	}
+
+	// send init actor msg to create the account actor
+	params := []interface{}{target}
+	targetIDAddrOpaque := ctx.Send(address.InitAddress, initactor.ExecMethodID, types.ZeroAttoFIL, params)
+	// cast response, interface{} -> address.Address
+	targetIDAddr = targetIDAddrOpaque.(address.Address)
+
+	// load actor
+	targetActor, err := ctx.rt.state.GetActor(context.Background(), targetIDAddr)
+	if err != nil {
+		panic("unreachable, exec failed to create the actor but returned successfully")
+	}
+
+	return targetActor, targetIDAddr
+}
+
+//
+// implement runtime.InvocationContext for invocationContext
+//
+
+var _ runtime.InvocationContext = (*invocationContext)(nil)
+
+// Runtime implements runtime.InvocationContext.
+func (ctx *invocationContext) Runtime() runtime.Runtime {
+	return ctx.rt
+}
+
+// Message implements runtime.InvocationContext.
+func (ctx *invocationContext) Message() runtime.MessageInfo {
+	return ctx.msg
+}
+
+// ValidateCaller implements runtime.InvocationContext.
+func (ctx *invocationContext) ValidateCaller(pattern runtime.CallerPattern) {
+	if ctx.isCallerValidated {
+		runtime.Abort("Method must validate caller identity exactly once")
+	}
+	if !pattern.IsMatch((*patternContext2)(ctx)) {
+		runtime.Abort("Method invoked by incorrect caller")
+	}
+	ctx.isCallerValidated = true
+}
+
+// StateHandle implements runtime.InvocationContext.
+func (ctx *invocationContext) StateHandle() runtime.ActorStateHandle {
+	return ctx.stateHandle
+}
+
+// LegacySend implements runtime.InvocationContext.
+func (ctx *invocationContext) LegacySend(to address.Address, method types.MethodID, value types.AttoFIL, params []interface{}) ([][]byte, uint8, error) {
+	panic("legacy code invoked")
+}
+
+// Send implements runtime.InvocationContext.
+func (ctx *invocationContext) Send(to address.Address, method types.MethodID, value types.AttoFIL, params []interface{}) interface{} {
+	// check if side-effects are allowed
+	if !ctx.allowSideEffects {
+		runtime.Abort("Calling Send() is not allowed during side-effet lock")
+	}
+
+	// prepare
+	// 1. alias fromActor
+	// 2. build internal message
+
+	// 1. fromActor = executing toActor
+	from := ctx.msg.to
+	fromActor := ctx.toActor
+
+	// 2. build internal message
+	encodedParams, err := abi.ToEncodedValues(params...)
+	if err != nil {
+		// Review: should this be a recoverable or unrecoverable panic?
+		panic("failed to encode params")
```

Comment:
I think we may have to do encoding in the actor, before Send() - I'll discuss elsewhere. But if we do it this way, this should abort rather than panic - it will be quite possible to pass an unencodable value.